### PR TITLE
Add broker controllers

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/src/controllers/MSIDRequestControllerFactory.m [hunk 1]
```diff
@@ -78,19 +104,23 @@ @implementation MSIDRequestControllerFactory
...
-        if (brokerInstalled)
+        if (@available(iOS 13.0, *))
         {
-            return [[MSIDBrokerInteractiveController alloc] initWithInteractiveRequestParameters:parameters
-                                                                            tokenRequestProvider:tokenRequestProvider
-                                                                              fallbackController:localController
-                                                                                           error:error];
+            __auto_type controller = [[MSIDBrokerExtensionInteractiveTokenRequestController alloc] initWithInteractiveRequestParameters:parameters
+                                                                                                                   tokenRequestProvider:tokenRequestProvider
+                                                                                                                     fallbackController:localController
```
Comment (oldalton) [inline IdentityCore/src/controllers/MSIDRequestControllerFactory.m]: should there also be a fallback controller for silent?

File: IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#import "MSIDInteractiveRequestParameters.h"
+#import "MSIDInteractiveTokenRequest.h"
+
+@interface MSIDBrokerExtensionInteractiveTokenRequestController ()
+
+@property (nonatomic) MSIDInteractiveTokenRequest *currentRequest;
+
+@end
+
+@implementation MSIDBrokerExtensionInteractiveTokenRequestController
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m]: this seems similar to the normal interactive controller. Should we reuse same class with different requests perhaps?

File: IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            
+            self.currentRequest = nil;
+            completionBlock(result, error);
+        };
+        
+        completionBlockWrapper(result, error);
+    }];
+}
+
+- (BOOL)canPerformRequest
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m]: previously we would check for presence of broker before initializing broker controller. Now, some of the checking is done before initialization and some after. Perhaps we should be consistent and always check for presence/capability of device before initialization a specific controller?

File: IdentityCore/src/controllers/broker/MSIDBrokerExtensionSilentTokenRequestController.m [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    if (!completionBlock)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, self.requestParameters, @"Passed nil completionBlock. End silent broker extension flow.");
+        return;
+    }
+
+    self.currentRequest = [self.tokenRequestProvider silentBrokerExtensionTokenRequestWithParameters:self.requestParameters
+                                                                                           forceRefresh:self.forceRefresh];
+
+    [self.currentRequest executeRequestWithCompletion:^(MSIDTokenResult *result, NSError *error)
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDBrokerExtensionSilentTokenRequestController.m]: same here, seems a bit similar to normal silent controller. Consider reusing code.

File: IdentityCore/src/controllers/broker/MSIDBrokerInteractiveController.m [hunk 5]
```diff
@@ -126,6 +126,36 @@ - (void)acquireToken:(MSIDRequestCompletionBlock)completionBlock
      }];
 }
 
+- (BOOL)canPerformRequest
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDBrokerInteractiveController.m]: I'd prefer to check for ability to use a particular controller before initializing a controller

File: IdentityCore/src/parameters/MSIDRequestParameters+Broker.m [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#import "MSIDAuthority.h"
+
+@implementation MSIDRequestParameters (Broker)
+
+- (BOOL)canUseBroker
+{
+#if TARGET_OS_IPHONE
+    if (self.requestType != MSIDRequestBrokeredType) return NO;
+    
+    if ([MSIDAppExtensionUtil isExecutingInAppExtension]) return NO;
```
Comment (oldalton) [inline IdentityCore/src/parameters/MSIDRequestParameters+Broker.m]: would this be still the case for the SSO extension?

File: IdentityCore/src/controllers/MSIDRequestControllerFactory.m [hunk 7]
```diff
@@ -28,20 +28,46 @@
 #if TARGET_OS_IPHONE
 #import "MSIDAppExtensionUtil.h"
 #import "MSIDBrokerInteractiveController.h"
+#import "MSIDBrokerExtensionInteractiveTokenRequestController.h"
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/controllers/MSIDRequestControllerFactory.m]: done

File: IdentityCore/src/controllers/MSIDRequestControllerFactory.m [hunk 8]
```diff
@@ -28,20 +28,46 @@
...
+                                                               error:(NSError **)error
 {
-    return [[MSIDSilentController alloc] initWithRequestParameters:parameters
-                                                      forceRefresh:forceRefresh
-                                              tokenRequestProvider:tokenRequestProvider
-                                                             error:error];
+    MSIDSilentController *brokerController;
+    
+#if TARGET_OS_IPHONE
+    if ([parameters canUseBroker])
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/controllers/MSIDRequestControllerFactory.m]: Renamed canUseBroker to shouldUseBroker. This checks if any broker (broker or extension) should be called according to the provided parameters.

File: IdentityCore/src/parameters/MSIDRequestParameters+Broker.m (hunk 6 above)
Comment (antrix1989) [reply IdentityCore/src/parameters/MSIDRequestParameters+Broker.m]: Addressed. Moved this check for the origin broker only.

File: IdentityCore/src/controllers/MSIDRequestControllerFactory.m (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/MSIDRequestControllerFactory.m]: We are calling silent flow with silent sso flow as a fallback.

File: IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m]: Moved to static method.

File: IdentityCore/src/controllers/broker/MSIDBrokerInteractiveController.m (hunk 5 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDBrokerInteractiveController.m]: done

File: IdentityCore/src/controllers/broker/MSIDBrokerExtensionSilentTokenRequestController.m (hunk 4 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDBrokerExtensionSilentTokenRequestController.m]: done

File: IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDBrokerExtensionInteractiveTokenRequestController.m]: done

File: IdentityCore/src/controllers/MSIDRequestControllerFactory.m [hunk 9]
```diff
@@ -28,20 +28,48 @@
...
+                                                                                                 error:error];
+            }
+        }
+    }
+#endif
+    
+    // TODO: Performance optimization: check account source.
+    // if (parameters.accountIdentifier.source == BROKER) return brokerController;
+    
+    __auto_type localController = [[MSIDSilentController alloc] initWithRequestParameters:parameters
```
Comment (oldalton) [inline IdentityCore/src/controllers/MSIDRequestControllerFactory.m]: shouldn't we check local cache first?

**Pull request merged**